### PR TITLE
[Feature] Charting Modularity

### PR DESCRIPTION
**Pull request by hjoaquim**
1. **Why**? (1-3 sentences or a bullet point list):

    - Atm is not possible to implement views in a given extension; this means that in order to implement those, one would need to touch the source code of the charting extension, which is not desireble and adds unnecessary friction.

2. **What**? (1-3 sentences or a bullet point list):

    - In this PR it's proposed an alternative implementation that extends the charting extension through the `pandas` accessors mechanics (already in place for extending the `OBBject` class).

3. **Impact** (1-2 sentences or a bullet point list):

    - TBD

4. **Testing Done**:

```python
from openbb import obb
res = obb.equity.price.historical("AAPL", chart=True)
res.show()
```

> the same but on the CLI

5. **Reviewer Notes** (optional):

    - Migrated all views to their own extension and deprecated `charting_router.py`

### CODE REVIEW

**Comment by piiq** [comment]
Does this change entail that openbb-etf will by default rely on the charting extension?

**Comment by hjoaquim** [comment]
> Does this change entail that openbb-etf will by default rely on the charting extension?

That's an oversight on me that were enthusiastic developing this and wasn't really considering dependencies tbh.

The `__init__.py` files should look like this instead:

```python
"""OpenBB ETF Extension."""

try:
    from openbb_charting import Charting  # type: ignore
    from openbb_etf import etf_views
except ImportError:
    pass

```

With this change anything else is needed, the `<extension>_views.py` will never be called so the imports pointing to the charting extension aren't a problem.

**Comment by deeleeramone** [comment]
So I was trying to checkout this branch and I wasn't able to import because of an AttributeError raised by PackageBuilder.

```
(cli) Danglewoods-MacBook-Pro:openbb_platform danglewood$ python -c "import openbb;openbb.build()"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/danglewood/GitHub/OpenBBTerminal/openbb_platform/openbb/__init__.py", line 39, in <module>
    _PackageBuilder(_this_dir).auto_build()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/site-packages/openbb_core/app/static/package_builder.py", line 91, in __init__
    self.route_map = PathHandler.build_route_map()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/site-packages/openbb_core/app/static/package_builder.py", line 1279, in build_route_map
    router = RouterLoader.from_extensions()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/site-packages/openbb_core/app/router.py", line 647, in from_extensions
    for name, entry in ExtensionLoader().core_objects.items():  # type: ignore[attr-defined]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/site-packages/openbb_core/app/extension_loader.py", line 110, in core_objects
    self._core_objects = self._load_entry_points(
                         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/site-packages/openbb_core/app/extension_loader.py", line 175, in _load_entry_points
    return func[group](entry_points_)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/site-packages/openbb_core/app/extension_loader.py", line 151, in load_core
    return {
           ^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/site-packages/openbb_core/app/extension_loader.py", line 152, in <dictcomp>
    ep.name: entry for ep in eps if isinstance((entry := ep.load()), Router)
                                                         ^^^^^^^^^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/site-packages/importlib_metadata/__init__.py", line 211, in load
    module = import_module(match.group('module'))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danglewood/miniconda3/miniconda3/envs/cli/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danglewood/GitHub/OpenBBTerminal/openbb_platform/extensions/crypto/openbb_crypto/__init__.py", line 6, in <module>
    from openbb_crypto import crypto_views
  File "/Users/danglewood/GitHub/OpenBBTerminal/openbb_platform/extensions/crypto/openbb_crypto/crypto_views.py", line 16, in <module>
    @ext.charting_accessor
     ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Extension' object has no attribute 'charting_accessor'
```

**Comment by hjoaquim** [comment]
> So I was trying to checkout this branch and I wasn't able to import because of an AttributeError raised by PackageBuilder.

Maybe try in a fresh env? Works perfectly on my end.

File: openbb_platform/extensions/crypto/openbb_crypto/crypto_views.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+class CryptoViews:
+    """Crypto Views."""
+
+    def __init__(self, charting: Charting) -> None:
+        """Initialize the crypto Views."""
+        self._charting = charting
+
+    @staticmethod
+    def crypto_price_historical(  # noqa: PLR0912
+        **kwargs,
```
[the remark this answers is not in the document]
Comment (hjoaquim) [reply openbb_platform/extensions/crypto/openbb_crypto/crypto_views.py:26]: These functions are not exposed directly, what would be the benefit? The implementation is the same as it was before on the `charting_router.py`; the charting extension calls a "generic" charting function to be executed.

There is no need to install anything, the `init` file of each extension checks if the charting extension is installed and if yes, adds the views to the charting accessor!

**Comment by deeleeramone** [comment]
> > So I was trying to checkout this branch and I wasn't able to import because of an AttributeError raised by PackageBuilder.
> 
> Maybe try in a fresh env? Works perfectly on my end.

This does not go as expected, all the lock files are downgrading the versions. :(

**Comment by hjoaquim** [comment]
> This does not go as expected, all the lock files are downgrading the versions. :(

Wdym? This PR doesn't change any dependencies or files related with them. Maybe I'm missing something?

**Comment by deeleeramone** [comment]
> Wdym? This PR doesn't change any dependencies or files related with them. Maybe I'm missing something?

That's this: https://github.com/OpenBB-finance/OpenBBTerminal/compare/660846c9d6ea040fcce3a8cd4a0fa4cec8489433..fa21f0493d39183aafcb0e154147a0746788f8cd

**Comment by hjoaquim** [comment]
> > Wdym? This PR doesn't change any dependencies or files related with them. Maybe I'm missing something?
> 
> That's this: https://github.com/OpenBB-finance/OpenBBTerminal/compare/660846c9d6ea040fcce3a8cd4a0fa4cec8489433..fa21f0493d39183aafcb0e154147a0746788f8cd

Was reverted, should be good now!

**Comment by deeleeramone** [comment]
Now I've got it going, I find this confusing. What is the expected use for all of these view accessors? They seem largely irrelevant to the returned OBBject and it's unclear what you are supposed to do with them.

Screenshot 2024-06-06 at 9 19 25 AM

**Review by piiq — APPROVED** [review]

**Comment by hjoaquim** [comment]
> Now I've got it going, I find this confusing. What is the expected use for all of these view accessors? They seem largely irrelevant to the returned OBBject and it's unclear what you are supposed to do with them.

The charting functions are not to be used by themselves; as we were also not using them when it was implemented on the `charting_router.py`, instead, it's the charting extension job to "connect the dots". For the end user it should be exactly the same as was before, i.e., you'd just need to use `chart=True` when available!

This is simply a different way of organizing the code, everything should still be the same UX-wise.
We can make the charting acessors private so it's explicit the user shouldn't use them (although, ultimately, you can bc it's all static methods!)

**Comment by deeleeramone** [comment]
> We can make the charting acessors private so it's explicit the user shouldn't use them (although, ultimately, you can bc it's all static methods!)

I think this will provide more clarity to users as these appear like something intended for use.

File: openbb_platform/extensions/crypto/openbb_crypto/crypto_views.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (deeleeramone) [reply openbb_platform/extensions/crypto/openbb_crypto/crypto_views.py:26]: Yes, making those methods private will get rid of most of the issue on the user side.

**Comment by hjoaquim** [comment]
> > We can make the charting acessors private so it's explicit the user shouldn't use them (although, ultimately, you can bc it's all static methods!)
> 
> I think this will provide more clarity to users as these appear like something intended for use.

I'm good with that! We can expand/improve this if/when we agree it's the way to go!

**Comment by IgorWounds** [comment]
@hjoaquim I like this!

**Review by montezdesousa — APPROVED** [review]

**Pull request merged**